### PR TITLE
Clean up code by removing support for LLVM older than 4

### PR DESCRIPTION
**Pull request by dmk42**
Supporting old versions of LLVM caused considerable clutter (over 500 lines worth).  This change removes the old `#ifdef`s.  Where reasonable, it also further consolidates the code once the conditionals are removed.

This implements part of #6993 .

While testing this, I ran up against `ClangInfo` being declared as a struct in one place and a class in another.  To get it to compile with the latest Mac Clang, I had to make it consistently a struct.

The following tests were run with both LLVM 4 and LLVM 5:
Passed full local testing with `--llvm` on linux64.
Passed the hellos with `--llvm`, with and without `--fast`, plus test/llvm on a Mac building LLVM with clang.

### CODE REVIEW

File: compiler/include/llvmDebug.h [hunk 1]
```diff
@@ -33,20 +33,14 @@
...
-#endif
 
 #include <vector>
 
 //#define DW_LANG_chapel (llvm::dwarf::DW_LANG_lo_user+37)
 #define DW_LANG_chapel llvm::dwarf::DW_LANG_lo_user
 
-#if HAVE_LLVM_VER >= 37
 // This should be renamed to DI versions
 #define LLVM_DITYPE llvm::DIType*
```
Comment (mppf) [inline compiler/include/llvmDebug.h]: I recommend we remove these #defines and static inline functions and replace their uses manually. They are a source of clutter / confusion. The only reason they were there was to support different LLVM versions.

File: compiler/include/llvmDebug.h [hunk 2]
```diff
@@ -85,62 +79,9 @@ static inline LLVM_DIVARIABLE toDIVARIABLE(llvm::MDNode* N)
...
-
-
-#endif
-
-
-
 // LLVM 4.0 added DIGlobalVariableExpression but previously
 // it was basically stored in DIGlobalVariable
-#if HAVE_LLVM_VER >= 40
 #define LLVM_DIGLOBALVARIABLEEXPRESSION llvm::DIGlobalVariableExpression*
```
Comment (mppf) [inline compiler/include/llvmDebug.h]: Let's remove the macro and replace its uses with the RHS.

File: compiler/include/llvmUtil.h [hunk 3]
```diff
@@ -46,39 +46,16 @@ static inline bool llvm_fn_param_has_attr(llvm::Function* f, unsigned idx, llvm:
...
 #error LLVM version is too old for this version of Chapel
 
 #endif
 
-// PassManager also changed names..
-#if HAVE_LLVM_VER >= 37
 #include "llvm/IR/LegacyPassManager.h"
 #define LEGACY_FUNCTION_PASS_MANAGER llvm::legacy::FunctionPassManager
 #define LEGACY_PASS_MANAGER llvm::legacy::PassManagerBase
 #define LEGACY_MODULE_PASS_MANAGER llvm::legacy::PassManager
```
Comment (mppf) [inline compiler/include/llvmUtil.h]: Let's remove the macros

File: compiler/include/llvmUtil.h [hunk 4]
```diff
@@ -103,7 +80,6 @@ uint64_t getTypeFieldNext(const llvm::DataLayout& layout, llvm::Type* ty, uint64
 
 
 // And create a type for a metadata operand 
-#if HAVE_LLVM_VER >= 36
 #define LLVM_METADATA_OPERAND_TYPE llvm::Metadata
```
Comment (mppf) [inline compiler/include/llvmUtil.h]: this one too

File: compiler/include/llvmUtil.h [hunk 5]
```diff
@@ -103,7 +80,6 @@ uint64_t getTypeFieldNext(const llvm::DataLayout& layout, llvm::Type* ty, uint64
...
 static inline llvm::ConstantAsMetadata* llvm_constant_as_metadata(llvm::Constant* C)
```
Comment (mppf) [inline compiler/include/llvmUtil.h]: and this and the following function

File: compiler/include/llvmUtil.h [hunk 6]
```diff
@@ -114,24 +90,9 @@ static inline
...
-
-#endif
 
 // LLVMs after 3.5 require C++11 support
-#if HAVE_LLVM_VER >= 35
 #define HAVE_LLVM_CXX11 1
-#endif
 
 #ifdef HAVE_LLVM_CXX11
 #define LLVM_CXX_OVERRIDE override
```
Comment (mppf) [inline compiler/include/llvmUtil.h]: We can remove this macro as well, along with HAVE_LLVM_CXX11

File: compiler/util/llvmDebug.cpp [hunk 7]
```diff
@@ -61,11 +61,7 @@ for more information on LLVM debug information.
 */
 
 // smooth changes in zero DIFlags
-#if HAVE_LLVM_VER >= 40
 #define FLAG_ZERO llvm::DINode::FlagZero
```
Comment (mppf) [inline compiler/util/llvmDebug.cpp]: Let's remove this macro

**Comment by dmk42** [comment]
Thanks @mppf !  I've made the suggested changes and retested.  Since I already have approval, I'll go ahead and merge so I can check the nightly tests tomorrow.

**Pull request merged**